sstables/sstable_set.hh

```diff
@@ -24,84 +24,207 @@
 #include "shared_sstable.hh"
 #include "dht/i_partitioner.hh"
 #include <seastar/core/shared_ptr.hh>
+#include <seastar/core/weak_ptr.hh>
 #include <vector>
 
 namespace sstables {
 
-class sstable_set_impl;
-class incremental_selector_impl;
+struct versioned_sstable_data {
+    shared_sstable sst;
+    uint64_t version_added;
+    std::optional<uint64_t> version_erased;
+};
+
+using versioned_sstable = seastar::lw_shared_ptr<versioned_sstable_data>;
+
+class incremental_selector_impl {
+public:
+    virtual ~incremental_selector_impl() {}
+    virtual std::tuple<dht::partition_range, std::vector<versioned_sstable>, dht::ring_position_view> select(const dht::ring_position_view&) = 0;
+};
+
+class sstable_set_impl {
+public:
+    virtual ~sstable_set_impl() {}
+    virtual std::vector<versioned_sstable> select(const dht::partition_range& range) const = 0;
+    virtual void insert(versioned_sstable sst) = 0;
+    virtual void erase(versioned_sstable sst) = 0;
+    virtual std::unique_ptr<incremental_selector_impl> make_incremental_selector() const = 0;
+};
 
 // Structure holds all sstables (a.k.a. fragments) that belong to same run identifier, which is an UUID.
 // SStables in that same run will not overlap with one another.
 class sstable_run {
-    sstable_list _all;
+    std::unordered_set<shared_sstable> _all;
 public:
+    sstable_run(std::unordered_set<shared_sstable>&&);
     void insert(shared_sstable sst);
     void erase(shared_sstable sst);
     // Data size of the whole run, meaning it's a sum of the data size of all its fragments.
     uint64_t data_size() const;
-    const sstable_list& all() const { return _all; }
+    const std::unordered_set<shared_sstable>& all() const { return _all; }
 };
 
-class sstable_set {
+class sstable_set_snapshot;
+class version_reference;
+
+class sstable_list {
```

Comment:
Actually, the sstable_list name isn't new - it was the return type of sstable_set::all() before the change too, but it was just an unordered_set then. I tried to explain it somewhat in the commit message, but I wasn't sure if something like it would be suitable for putting in the source code.